Running-Mastodon/Migration-guide.md

```diff
@@ -0,0 +1,79 @@
+Migration guide
+====
+
+Sometimes, for various reasons, you may want to migrate your Mastodon instance from one server to another. Fortunately this is not too difficult of a process, although it may result in some downtime for your users.
+
+**Note:** this guide was written with Ubuntu Server 16.04 in mind; your mileage may vary for other setups.
+
+What needs to be migrated
+---
+
+At a high level, you'll need to copy over the following:
+
+- The `~/live/public/system` directory, which contains user-uploaded images and videos
+- The Postgres database (using [pg\_dump](https://www.postgresql.org/docs/9.1/static/backup-dump.html))
+- The `~/live/.env.production` file, which contains server config and secrets
+
+Less crucially, you'll proably also want to copy the following for convenience:
+
+- The nginx config (under `/etc/nginx/sites-available/default`)
+- The systemd config files (`/etc/systemd/system/mastodon-*.service`), which may contain your server tweaks and customizations
+- The pgbouncer configuration under `/etc/pgbouncer` (if you're using it)
+
+Preliminary steps
+----
+
+All of the following steps can be taken without any downtime for your existing server.
+
+First, follow the [Production Guide](https://github.com/tootsuite/documentation/blob/master/Running-Mastodon/Production-guide.md) to set up a brand-new Mastodon server. However, don't set up a new Postgres database using `db:setup`; we're going to copy it over instead.
+
+### Dump and load Postgres
+
+Create an empty Postgres database using the `template0` database (which is useful when restoring a Postgres dump, [as described in the pg\_dump documentation](https://www.postgresql.org/docs/9.1/static/backup-dump.html#BACKUP-DUMP-RESTORE)). You'll want to run this as the `mastodon` user on your new system:
+
+```bash
+createdb -T template0 mastodon_production
+```
+
+Next, create a Postgres dump on the old system (again, as the `mastodon` user):
+
+```bash
+pg_dump mastodon_production > dump.sql
+```
+
+Copy the `dump.sql` file over, using `scp` or `rsync`. Then on the new system, run:
+
+```bash
+psql mastodon_production < dump.sql
```

Comment:
I usually use pg_restore, e.g.: <https://devcenter.heroku.com/articles/heroku-postgres-import-export#restore-to-local-database>

Reply:
Good call, this does seem more appropriate!

Reply:
Hm actually after some research, it seems `pg_restore` only works with the non-plaintext output of `pg_dump`. Also the script version [may be more robust](https://www.postgresql.org/docs/current/static/app-pgdump.html):

> Script files can be used to reconstruct the database even on other machines and other architectures; with some modifications, even on other SQL database products.

Maybe it's best to just keep the basic version, for simplicity's sake. :)